xwiki-commons-core/xwiki-commons-job/src/main/java/org/xwiki/job/internal/JobGroupPathLockTree.java

```diff
@@ -32,31 +38,104 @@
  * @version $Id$
  * @since 6.1M2
  */
+@Component(roles = JobGroupPathLockTree.class)
+@Singleton
 public class JobGroupPathLockTree
 {
-    private final Map<JobGroupPath, ReadWriteLock> tree = new ConcurrentHashMap<>();
+    @Inject
+    private GroupedJobInitializerManager groupedJobInitializerManager;
+
+    /**
+     * A specific concurrency implementation for managing Semaphore with Read/Write lock capabilities.
+     */
+    private class ReadWriteSemaphore
```

Comment:
Might worth putting that in its own file ideally somewhere in commons maybe since it's very generic.

Reply:
Might make sense. Not as a public API though but internal, right?

Reply:
Why not public ? So that we can ditch it if we find a lib that does this ?

Reply:
> So that we can ditch it if we find a lib that does this ?

yep.

Reply:
OK then good point.